src/sdam/monitor.rs

```diff
@@ -28,20 +29,33 @@ lazy_static! {
 /// Starts a monitoring thread associated with a given Server. A weak reference is used to ensure
 /// that the monitoring thread doesn't keep the server alive after it's been removed from the
 /// topology or the client has been dropped.
-pub(super) fn monitor_server(
-    mut conn: Connection,
-    server: Weak<Server>,
-    heartbeat_frequency: Option<Duration>,
-) {
+pub(super) fn monitor_server(address: StreamAddress, server: Weak<Server>, options: ClientOptions) {
     std::thread::spawn(move || {
         let mut server_type = ServerType::Unknown;
-        let heartbeat_frequency = heartbeat_frequency.unwrap_or(DEFAULT_HEARTBEAT_FREQUENCY);
+        let heartbeat_frequency = options
+            .heartbeat_freq
+            .unwrap_or(DEFAULT_HEARTBEAT_FREQUENCY);
+
+        let make_connection = || {
```

Comment:
I took a slightly different approach in node, and I wonder what your thoughts are as they may apply to Rust. 

Instead of passing a function around here, I just made the "connection" a member of the `Monitor` (in this case it might be a member of the `Server` since there is no `Monitor` type). Then I have a method called `checkServer` which either reuses the existing connection (known good), or if the existing connection is `null` then it makes a new connection. The added benefit is that, since the initial handshake of connection establishment _is_ itself a heartbeat, you can still call the whole process `checkServer`.

Reply:
I took a stab at rewriting this a bit to use a `Monitor` struct; let me know what you think of it

Reply:
nice, LGTM!